community/index/src/main/java/org/neo4j/index/gbptree/GBPTree.java

```diff
@@ -0,0 +1,481 @@
+     * be written in index header.
+     * If the index exists it will be opened and the {@link Layout} will be matched with the information
+     * in the header. At the very least {@link Layout#identifier()} will be matched, but also if the
+     * index has {@link Layout#writeMetaData(PageCursor)} additional meta data it will be
+     * {@link Layout#readMetaData(PageCursor)}.
+     *
+     * @param pageCache {@link PageCache} to use to map index file
+     * @param indexFile {@link File} containing the actual index
+     * @param tentativePageSize page size, i.e. tree node size. Must be less than or equal to that of the page cache.
+     * A pageSize of {@code 0} means to use whatever the page cache has (at creation)
+     * @param layout {@link Layout} to use in the tree, this must match the existing layout
+     * we're just opening the index
+     * @throws IOException on page cache error
+     */
+    public GBPTree( PageCache pageCache, File indexFile, Layout<KEY,VALUE> layout, int tentativePageSize )
+            throws IOException
+    {
+        this.layout = layout;
+        this.pagedFile = openOrCreate( pageCache, indexFile, tentativePageSize, layout );
+        this.bTreeNode = new TreeNode<>( pageSize, layout );
+        this.writer = new AtomicReference<>(
+                new SingleIndexWriter( new InternalTreeLogic<>( this, bTreeNode, layout ) ) );
+
+        if ( created )
+        {
+            // Initialize index root node to a leaf node.
+            try ( PageCursor cursor = pagedFile.io( rootId, PagedFile.PF_SHARED_WRITE_LOCK ) )
+            {
+                cursor.next();
+                bTreeNode.initializeLeaf( cursor, stableGeneration, unstableGeneration );
+            }
+        }
+    }
+
+    private PagedFile openOrCreate( PageCache pageCache, File indexFile,
+            int pageSizeForCreation, Layout<KEY,VALUE> layout ) throws IOException
+    {
+        try
+        {
+            PagedFile pagedFile = pageCache.map( indexFile, pageCache.pageSize() );
+            // This index already exists, verify the header with what we got passed into the constructor this time
+
+            try
+            {
+                // Read header
+                long layoutIdentifier;
+                int majorVersion;
+                int minorVersion;
+                try ( PageCursor metaCursor = openMetaPageCursor( pagedFile ) )
+                {
+                    do
+                    {
+                        pageSize = metaCursor.getInt();
+                        rootId = metaCursor.getLong();
+                        lastId = metaCursor.getLong();
+                        layoutIdentifier = metaCursor.getLong();
+                        majorVersion = metaCursor.getInt();
+                        minorVersion = metaCursor.getInt();
+                        layout.readMetaData( metaCursor );
+                    }
+                    while ( metaCursor.shouldRetry() );
```

Comment:
Not until we, if at all, start setting cursor exceptions in our write execution path. To my knowledge the cursor exception is completely user controlled and serves as a way for application code to transfer an exception through a "shouldRetry" barrier.